_posts/2019-12-18-package-lock.md

```diff
@@ -0,0 +1,101 @@
+---
+layout: post
+title: To lock or not to lock
+tags:
+- node.js
+status: publish
+category: node.js
+type: post
+published: true
+---
+
+![Package Lock Files](/images/2019/lock.png)
+
+A few years ago, [npm](https://npmjs.org) introduced the notion of a [`package-lock.json`](https://docs.npmjs.com/files/package-lock.json). The purpose of the file is to provide a manifest that calls out the exact version of every package in your tree, the last time `npm install` was run.  After running `npm install`, you're going to see a message like this:
+
+<pre><code>npm notice created a lockfile as package-lock.json. You should commit this file.
+</code></pre>
+
+While this advice is well intentioned...  it's not always true :) Let's talk about when you want to check this into source control, and when not to.
+
+## How package-lock.json works
+
+The mechanics of `package-lock.json` are simple enough:
+- When you run `npm install` the first time, npm automatically creates the file for you.
+- It has the exact version of every top level and transitive dependency in your `node_modules` folder.
+- When you update the version of a package in your `package.json` and run `npm install`, the `package-lock.json` file will get updated automatically.
```

Comment:
...or just install a new package. Actually, I would better say "when you install a new package", because people normally just do `npm i whatever` and don't bother editing `package.json` manually, I guess.